# TIL

1일1커밋 무사고: 245일차

06:25

## todo

- [ ] flash card
  - [ ] 컨벤션 문서 추가
    - [ ] PR은 당일 merge 불가능만 유지하기 컨벤션 추가하기
    - [ ] 1일 1PR만 merge 가능한 것은 해제하기
  - [ ] backend 작업
    - [ ] deno deploy docker image 찾기
    - [ ] 배포환경 바꿔보기
    - [x] 계정 삭제 요청
  - [x] frontend
    - [x] card side useReducer로 리팩토링
    - [x] 에러 바운더리
    - [x] 에러 이후 Retry 추가
- [ ] 코테
  - 간단하게 3문제
- [ ] 이력서
  - [x] 1뽀모도로
  - [x] 1뽀모도로
  - [ ] 1뽀모도로
  - [ ] 1뽀모도로
  - [ ] 1뽀모도로
  - [ ] 1뽀모도로
- 입사지원
  - [ ] 퀀다과외
    - [ ] 자소서 작성하기
  - [ ] [레티튜(Letitu)](https://www.wanted.co.kr/wd/125018)
    - [ ] 리덕스 활용경험 자소서에 추가
    - [ ] action cost 감소 이력서에 추가
  - [x] 와이아웃
  - [x] 카르도
  - [x] 익스토리
  - [x] 헥슬란트
  - [x] 브이씨에이테크놀로지
  - [x] 이력서 수정 leading edge debounce 정정
- [ ] 최종졸업학교의 성적증명서 출력

## 에러 바운더리

https://www.youtube.com/watch?v=0LpLxEUephc

```tsx
import { Component, ErrorInfo, ReactNode } from 'react';

interface Props {
  children?: ReactNode;
  fallback: ReactNode;
}

interface State {
  hasError: boolean;
}

export class ErrorBoundary extends Component<Props, State> {
  public state: State = {
    hasError: false,
  };

  public static getDerivedStateFromError(_: Error): State {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.error('Uncaught error:', error, errorInfo);
  }

  public render() {
    if (this.state.hasError) {
      return this.props.fallback;
    }

    return this.props.children;
  }
}
```

```tsx title="Deck.tsx"
function Deck() {
  return (
    <ErrorBoundary fallback={<h2>something went wrong???</h2>}>
      <DeckList />
    </ErrorBoundary>
  );
}

export default Deck;
```

```tsx
import { useCards } from '@/hooks';

export function DeckList() {
  const { cards } = useCards();

  const IntervalArray: { title: string; deck: Card[] }[] = useMemo(() => {
    // ... 생략
  }, [cards]);

  return (
    <DeckListContainer>
      {IntervalArray.map((deckItem, idx) => (
        <DeckItem title={deckItem.title} cards={deckItem.deck} key={idx} />
      ))}
    </DeckListContainer>
  );
}
```

여기 컴포넌트에서 통신을 처리합니다. 통신을 처리하면 항상 실패가능성이 있습니다. 리액트의 경우 상위 컴포넌트에서 ErrorBoundary로 처리할 수 있습니다.

하지만 문제가 있습니다. 자식컴포넌트에서 Error에 대한 정보가 있습니다. 부모의 컴포넌트가 어떻게 Error에 대한 정보를 접근하고 표시할 수 있는가? 다른 문제도 있습니다. 에러를 화면에 표시하는 것은 View에 해당하는 로직입니다. 에러에 대한 정보는 보통 통신과 많이 하고 관심사가 Controller에 가깝습니다. 즉 필요한 것은 조합입니다. 에러마다 다른 컴포넌트에 에러를 보여줘야 합니다.

결론: 그냥 라이브러리 설치합시다.

https://github.com/bvaughn/react-error-boundary

```sh
yarn add react-error-boundary
```

```tsx
import { EmptyCards } from '@/Components';
import { NowDeck } from './subcomponents';

function Cards() {
  return (
    <ErrorBoundary FallbackComponent={EmptyCards}>
      <NowDeck />
    </ErrorBoundary>
  );
}

export default Cards;
```

```tsx
import { NoCardContainer, NoCardMessage } from './EmptyCards.style';

type EmptyCardsProps = {
  error?: { success: boolean; msg: string };
};

export function EmptyCards({ error }: EmptyCardsProps) {
  return (
    <NoCardContainer>
      <NoCardMessage>{error ? error.msg : '카드가 없습니다.'}</NoCardMessage>
    </NoCardContainer>
  );
}
```

이렇게 처리하면 동적으로 Error fallback에 error 객체를 대입할 수 있게 됩니다.

### retry 추가

```tsx
import { ErrorBoundary } from 'react-error-boundary';
import { NowDeck } from './subcomponents';
import { useQueryErrorResetBoundary } from '@tanstack/react-query';

function Cards() {
  const { reset } = useQueryErrorResetBoundary();

  return (
    <ErrorBoundary
      onReset={reset}
      fallbackRender={({ resetErrorBoundary, error }) => (
        <EmptyCards error={error} resetErrorBoundary={resetErrorBoundary} />
      )}
    >
      <NowDeck />
    </ErrorBoundary>
  );
}

export default Cards;
```

재요청 버튼이 처리할 함수는 resetErrorBoundary입니다. 라이브러리에서 제공해줍니다.

하지만 설정할 때 `useQueryErrorResetBoundary`에서 `reset`을 `ErrorBoundary`에 대입해야 합니다.

```tsx title="EmptyCards.tsx"
import { Button } from '..';
import { ErrorCardMessage, NoCardMessage } from './EmptyCards.style';

type EmptyCardsProps = {
  error: ErrorResponse;
  resetErrorBoundary: (...args: any[]) => void;
};

export function EmptyCards({ error, resetErrorBoundary }: EmptyCardsProps) {
  return (
    <ErrorCardMessage>
      <NoCardMessage>{error.msg}</NoCardMessage>
      <Button onClick={resetErrorBoundary}>retry</Button>
    </ErrorCardMessage>
  );
}
```

이렇게 적용하면 됩니다. resetErrorBoundary을 handler로 props로 받아 원하는 Component 이벤트에 주입하면 됩니다.

suspense 적용할까 고민했지만 추가 제어가 조금 어려웠습니다.

어려운점은 공식문서 예시가 이상합니다.

https://tanstack.com/query/latest/docs/react/examples/react/suspense

codesandbox에서 VM이 실패합니다.

예시를 보고 이해할 수 있으면 적용해볼 수 있었을 것같습니다.

## nest.js

https://youtu.be/GHTA143_b-s?t=932

```ts title="auth.controller.ts"
import { Controller } from '@nestjs/common';

@Controller()
export class AuthController {}
```

```ts title="auth.service.ts"
import { Injectable } from '@nestjs/common';

@Injectable({})
export class AuthService {}
```

```ts title="auth.module.ts"
import { Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';

@Module({
  imports: [],
  controllers: [AuthController],
  providers: [AuthService],
})
export class AuthModule {}
```

---

```ts
import { Controller, Post } from '@nestjs/common';
import { AuthService } from './auth.service';

@Controller('auth')
export class AuthController {
  constructor(private authService: AuthService) {}

  @Post('signup')
  signup() {
    return 'sign up';
  }

  @Post('signin')
  signin() {
    return 'sign in';
  }
}
```

```
http://localhost:3000/auth/signup
```

post 요청을 하면 sign up 응답을 확인할 수 있습니다.

지금은 문자열인데 응답을 객체같은 참조형으로 응답해도 알아서 json으로 변환합니다.

```ts
import { Controller, Post } from '@nestjs/common';
import { AuthService } from './auth.service';

@Controller('auth')
export class AuthController {
  constructor(private authService: AuthService) {}

  @Post('signup')
  signup() {
    return { mas: 'good' };
  }

  @Post('signin')
  signin() {
    return 'sign in';
  }
}
```

요청응답처리는 controller가 처리합니다. 비즈니스 로직은 service에서 처리합니다.

```ts title="auth.service.ts"
import { Injectable } from '@nestjs/common';

@Injectable({})
export class AuthService {
  login() {
    return { msg: 'good' };
  }

  signup() {
    return { msg: 'fuiyoh' };
  }
}
```

```ts title="auth.controller.ts"
import { Controller, Post } from '@nestjs/common';
import { AuthService } from './auth.service';

@Controller('auth')
export class AuthController {
  constructor(private authService: AuthService) {}

  @Post('signup')
  signup() {
    return this.authService.signup();
  }

  @Post('signin')
  signin() {
    return this.authService.login();
  }
}
```

이렇게 하면 요청을 처리하는 객체와 비즈니스 로직을 처리하는 관심사가 분리됩니다.

요청과 비즈니스로직을 분리해서 비즈니스로직에 더 집중하는 전략입니다.

https://youtu.be/GHTA143_b-s?t=1579
